python/pyspark/sql/types.py

```diff
@@ -368,8 +367,49 @@ def __init__(self, fields):
         >>> struct1 == struct2
         False
         """
-        assert all(isinstance(f, DataType) for f in fields), "fields should be a list of DataType"
-        self.fields = fields
+        if not fields:
+            self.fields = []
+        else:
+            self.fields = fields
+            assert all(isinstance(f, StructField) for f in fields),\
+                "fields should be a list of StructField"
+
+    def add(self, name_or_struct_field, data_type=NullType(), nullable=True, metadata=None):
```

Comment:
What's the use cases that we should have StructType without specifying the dataType of each column?

In `createDataFrame`, if a schema of StructType is provided, it will not try to infer the data types, so it does not work with StructType with NoneType in it.